lib/Sema/DerivedConformanceComparable.cpp

```diff
@@ -0,0 +1,411 @@
+    generatedIdentifier = C.Id_LessThanOperator;
+  } else {
+    assert(selfIfaceTy->getEnumOrBoundGenericEnum());
+    generatedIdentifier = C.Id_derived_enum_less_than;
+  }
+
+  DeclName name(C, generatedIdentifier, params);
+  auto comparableDecl =
+    FuncDecl::create(C, /*StaticLoc=*/SourceLoc(),
+                     StaticSpellingKind::KeywordStatic,
+                     /*FuncLoc=*/SourceLoc(), name, /*NameLoc=*/SourceLoc(),
+                     /*Throws=*/false, /*ThrowsLoc=*/SourceLoc(),
+                     /*GenericParams=*/nullptr,
+                     params,
+                     TypeLoc::withoutLoc(boolTy),
+                     parentDC);
+  comparableDecl->setImplicit();
+  comparableDecl->setUserAccessible(false);
+
+  // Add the @_implements(Comparable, < (_:_:)) attribute
+  if (generatedIdentifier != C.Id_LessThanOperator) {
+    auto comparable = C.getProtocol(KnownProtocolKind::Comparable);
+    auto comparableType = comparable->getDeclaredType();
+    auto comparableTypeLoc = TypeLoc::withoutLoc(comparableType);
+    SmallVector<Identifier, 2> argumentLabels = { Identifier(), Identifier() };
+    auto comparableDeclName = DeclName(C, DeclBaseName(C.Id_LessThanOperator),
+                                   argumentLabels);
+    comparableDecl->getAttrs().add(new (C) ImplementsAttr(SourceLoc(),
+                                                          SourceRange(),
+                                                          comparableTypeLoc,
+                                                          comparableDeclName,
+                                                          DeclNameLoc()));
+  }
+
+  if (!C.getLessThanIntDecl()) {
+    derived.ConformanceDecl->diagnose(diag::no_less_than_overload_for_int);
+    return nullptr;
+  }
+
+  comparableDecl->setBodySynthesizer(bodySynthesizer);
+
+  comparableDecl->copyFormalAccessFrom(derived.Nominal, /*sourceIsParentContext*/ true);
+
+  // Add the operator to the parent scope.
+  derived.addMembersToConformanceContext({comparableDecl});
+
+  return comparableDecl;
+}
+
+bool 
+DerivedConformance::canDeriveComparable(DeclContext *context, NominalTypeDecl *declaration) {
+  // The type must be an enum.
+  if (EnumDecl *const enumeration = dyn_cast<EnumDecl>(declaration)) {
+    // The cases must not have non-comparable associated values or raw backing
+    auto comparable = context->getASTContext().getProtocol(KnownProtocolKind::Comparable);
+    return allAssociatedValuesConformToProtocol(context, enumeration, comparable) && !enumeration->hasRawType();
+  } else {
+    return false;
+  }
+}
+
+ValueDecl *DerivedConformance::deriveComparable(ValueDecl *requirement) {
+  if (checkAndDiagnoseDisallowedContext(requirement))
+    return nullptr;
+  // Build the necessary decl.
+  if (requirement->getBaseName() == "<") {
+    if (EnumDecl const *const enumeration = dyn_cast<EnumDecl>(this->Nominal)) {
+      auto bodySynthesizer = !enumeration->hasCases() 
```

Comment:
I would really prefer if this were broken out into a lambda with proper if guards.  I find nested ternary form to be completely unreadable.

Reply:
broke it into explicit if else statements